REGRESSION (268038@main): :not(:has(:not(foo))) misclassified as scope breaking

anttijk opened the pull request:
#### 4ff0fdab94ceae40ad1582477eac1b7800a11ba3
<pre>
REGRESSION (268038@main): :not(:has(:not(foo))) misclassified as scope breaking
<a href="https://bugs.webkit.org/show_bug.cgi?id=267628">https://bugs.webkit.org/show_bug.cgi?id=267628</a>
<a href="https://rdar.apple.com/120492012">rdar://120492012</a>

Reviewed by Cameron McCormack.

Seen in <a href="https://tc39.es/ecma262/#sec-completion-record-specification-type">https://tc39.es/ecma262/#sec-completion-record-specification-type</a>

* LayoutTests/fast/selectors/has-scope-breaking-classification-expected.txt: Added.
* LayoutTests/fast/selectors/has-scope-breaking-classification.html: Added.
* Source/WebCore/style/ChildChangeInvalidation.cpp:
(WebCore::Style::ChildChangeInvalidation::invalidateForChangeOutsideHasScope):
* Source/WebCore/style/RuleFeature.cpp:
(WebCore::Style::RuleFeatureSet::recursivelyCollectFeaturesFromSelector):

:not() outside :has() would still put us into CanBreakScope::Yes mode.
Fix by only setting CanBreakScope::Yes if we are actually inside :has().

* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::hasScopeBreakingHasSelectors const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Add testing support.

Canonical link: <a href="https://commits.webkit.org/273177@main">https://commits.webkit.org/273177@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36e3c58dcebd17dd2d4aa8f579696960eb4d104e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34574 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13398 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36581 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37264 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/31244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35716 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15796 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10505 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30219 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35099 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11338 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30812 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9904 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/9993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30877 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38542 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31400 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/31193 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36050 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10106 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7998 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34006 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11928 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/30502 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7942 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/10658 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10977 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/selectors/has-scope-breaking-classification.html:
```diff
@@ -0,0 +1,52 @@
+    log.textContent += selector + (isScopeBreaking ? " IS scope breaking\n" : " is not scope breaking\n");
+    style.remove();
+}
+
```
Can you add the `:has(+ :not(.class))` case from the other bug?

anttijk committed:
REGRESSION (268038@main): :not(:has(:not(foo))) misclassified as scope breaking
https://bugs.webkit.org/show_bug.cgi?id=267628
rdar://120492012

Reviewed by Cameron McCormack.

Seen in https://tc39.es/ecma262/#sec-completion-record-specification-type

* LayoutTests/fast/selectors/has-scope-breaking-classification-expected.txt: Added.
* LayoutTests/fast/selectors/has-scope-breaking-classification.html: Added.
* Source/WebCore/style/ChildChangeInvalidation.cpp:
(WebCore::Style::ChildChangeInvalidation::invalidateForChangeOutsideHasScope):
* Source/WebCore/style/RuleFeature.cpp:
(WebCore::Style::RuleFeatureSet::recursivelyCollectFeaturesFromSelector):

:not() outside :has() would still put us into CanBreakScope::Yes mode.
Fix by only setting CanBreakScope::Yes if we are actually inside :has().

* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::hasScopeBreakingHasSelectors const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Add testing support.

Canonical link: https://commits.webkit.org/273177@main
